File: game_logic.py
'''
/*********************/
/*  game_logic.py    */
/*   Version 1.0     */
/*    2025/04/27     */
/*********************/
'''
import numpy as np
from mod_config import cfg, game_cfg as gcfg, planet_cfg as pcfg, \
    lander_cfg as lcfg


class GameLogic:
    def __init__(self, no_print=False):
        self.reset()
        # Constants from config
        self.landing_pad_center_x = gcfg.lpad_x1 + gcfg.lpad_width / 2
        self.landing_pad_y = gcfg.pad_y1  # Using top boundary as reference
        self.no_print = no_print

    def p(self, text):
        if self.no_print:
            return
        print(text)

    def reset(self):
        """Resets the game state to the initial configuration."""
        self.x = gcfg.x0[0]
        self.y = gcfg.x0[1]
        self.vx = gcfg.v0[0]
        self.vy = gcfg.v0[1]
        # Initial acceleration includes gravity
        self.ax = gcfg.a0[0]
        self.ay = gcfg.a0[1] + pcfg.g  # Gravity is always acting
        self.fuel = float(lcfg.max_fuel)
        self.landed = False
        self.crashed = False
        self.landed_successfully = False
        self.out_of_bounds = False
        self.time_step = 0
        self.action_log = []  # To track actions if needed

    def _apply_action(self, action: int):
        """Applies the chosen action (thrust).
        0: Noop, 1: Up, 2: Left, 3: Right"""
        self.action_log.append(action)  # Log action
        if self.fuel <= 0:
            return  # No thrust if out of fuel

        if action == 1:  # Thrust Up
            # Apply vertical thrust impulse (counteracting gravity)
            self.vy -= 0.3  # Instantaneous change, adjust as needed
            self.fuel -= 1
        elif action == 2:  # Thrust Left
            self.vx -= 0.05
            self.fuel -= 0.5
        elif action == 3:  # Thrust Right
            self.vx += 0.05
            self.fuel -= 0.5
        # Action 0 is Noop, do nothing

        # Clamp fuel to zero
        self.fuel = max(0, self.fuel)

    def _update_physics(self):
        """Updates the lander's position and velocity based on physics."""
        # Update velocity with acceleration (gravity is already in ay)
        # self.vx += self.ax # No horizontal acceleration
        self.vy += self.ay

        # Apply friction
        self.vx *= (1 - pcfg.mu_x)
        self.vy *= (1 - pcfg.mu_y)  # Air resistance/friction

        # Update position
        self.x += self.vx
        self.y += self.vy

        # Check boundaries (simple check for now)
        if self.x < 0 or self.x + lcfg.width > cfg.width:
            self.vx *= -0.5  # Bounce off sides slightly
            self.x = np.clip(self.x, 0, cfg.width - lcfg.width)
            # self.crashed = True # Or just bounce?
            # self.out_of_bounds = True

        # Prevent going below ground level before landing check
        ground_level = cfg.height - lcfg.height - gcfg.terrain_y
        if self.y > ground_level:
            self.y = ground_level
            # Don't zero velocity here yet, check landing conditions first

    def _check_landing_crash(self):
        """Checks if the lander has landed or crashed.
        Only performs check after first step."""
        # Only check for landing/crash after the first time step
        # to avoid immediate crash on start
        if self.time_step == 0:
            return

        ground_level = cfg.height - lcfg.height - gcfg.terrain_y
        if self.y >= ground_level:
            self.y = ground_level  # Ensure it rests on the ground
            self.landed = True

            on_landing_pad = (self.x >= gcfg.lpad_x1 and
                              self.x + lcfg.width <= (gcfg.lpad_x1
                                                      + gcfg.lpad_width))
            safe_speed = (abs(self.vx) < abs(gcfg.max_vx) and
                          abs(self.vy) < abs(gcfg.max_vy))

            if on_landing_pad and safe_speed:
                self.landed_successfully = True
                self.vx = 0  # Stop movement on successful landing
                self.vy = 0
                self.p("GameLogic: Successful landing!")
            else:
                self.crashed = True
                self.landed_successfully = False
                # Stop movement on crash
                self.vx = 0
                self.vy = 0
                if not on_landing_pad:
                    self.p("GameLogic: Crashed - outside landing area!")
                if abs(self.vx) >= abs(gcfg.max_vx):
                    self.p("GameLogic: Crashed - excessive horizontal speed! "
                           f"(|{self.vx:.2f}| >= {gcfg.max_vx})")
                if abs(self.vy) >= abs(gcfg.max_vy):
                    self.p("GameLogic: Crashed - excessive vertical speed! ("
                           f"|{self.vy:.2f}| >= {gcfg.max_vy})")

    def update(self, action: int):
        """Performs one time step of the game simulation.
           Args: action (int): The action to apply
                               (0: Noop, 1: Up, 2: Left, 3: Right).
           Returns: tuple: (state, reward, done)
                      state (np.array): The current state vector for the NN.
                      reward (float): The reward obtained in this step.
                      done (bool): Whether the episode has ended.
        """
        if self.is_done():
            return self.get_state(), 0.0, True

        self._apply_action(action)
        self._update_physics()
        self._check_landing_crash()

        self.time_step += 1

        done = self.is_done()
        reward = self._calculate_reward(done)
        state = self.get_state()

        return state, reward, done

    def is_done(self) -> bool:
        """Checks if the game episode has finished."""
        # return self.landed or self.crashed or self.fuel <= 0
        # or self.out_of_bounds
        # For now, it ends only on landing/crash or add a time limit later
        return self.landed or self.crashed

    def _calculate_reward(self, done: bool) -> float:
        """Calculates the reward for the current state/action.
           Placeholder implementation. Needs refinement for RL.
        """
        reward = 0.0

        # Penalty for fuel consumption
        # reward -= 0.1 # Small penalty per step?
        # reward -= (self.action_log[-1] > 0) * 0.05 # Penalty for using fuel

        # Shaping reward: encourage getting closer to the pad
        dist_x = self.x - self.landing_pad_center_x
        dist_y = self.y - self.landing_pad_y  # Target is above ground
        reward -= (abs(dist_x) + abs(dist_y)) * 0.001  # Penalize distance

        if done:
            if self.landed_successfully:
                reward += 100.0  # Large reward for success
                reward -= self.time_step * 0.1  # Penalize taking too long
            elif self.crashed:
                reward -= 100.0  # Large penalty for crashing
            # elif self.fuel <= 0 and not self.landed:
            #     reward -= 50.0 # Penalty for running out of fuel mid-air
        return reward

    def get_state(self) -> np.ndarray:
        """Returns the current state vector for the NN."""
        # Calculate distance to target landing pad center
        dist_target_x = self.x - self.landing_pad_center_x
        dist_target_y = self.y - self.landing_pad_y  # Target Y is pad top

        # Normalize? For now, use raw values.
        # Normalization depends on the NN architecture and training process.
        state = np.array([
            self.x / cfg.width,             # Normalized X position
            self.y / cfg.height,            # Normalized Y position
            # Normalized Vx (relative to max safe speed)
            self.vx / gcfg.max_vx,
            # Normalized Vy (relative to max safe speed)
            self.vy / gcfg.max_vy,
            self.fuel / lcfg.max_fuel,      # Normalized Fuel
            dist_target_x / cfg.width,      # Normalized distance X
            dist_target_y / cfg.height,     # Normalized distance Y
            # self.time_step / MAX_STEPS ? # Optional: Normalized
        ], dtype=float)
        # Ensure vx/vy normalization doesn't cause issues
        # if max speeds are very small or zero
        # Clip normalized velocity to avoid extremes
        state[2] = np.clip(state[2], -5, 5)
        state[3] = np.clip(state[3], -5, 5)

        return state

    def get_render_info(self) -> dict:
        """Returns information needed for rendering."""
        return {
            "x": self.x,
            "y": self.y,
            "vx": self.vx,
            "vy": self.vy,
            "fuel": self.fuel,
            "last_action": self.action_log[-1] if self.action_log else 0,
            "landed": self.landed,
            "crashed": self.crashed,
            "landed_successfully": self.landed_successfully
        }


if __name__ == '__main__':
    # Example usage (optional)
    logic = GameLogic()
    state = logic.get_state()
    print("Initial State:", state)
    done = False
    step = 0
    total_reward = 0
    # Simulate a few steps with random actions
    while not done and step < 500:
        action = np.random.randint(0, 4)  # Random action
        state, reward, done = logic.update(action)
        total_reward += reward
        # print(f"Step {step}: Action={action}, Reward={reward:.2f}, "
        #      f"Done={done}")
        # print(f" State: x={logic.x:.1f}, y={logic.y:.1f}, "
        #      f"vx={logic.vx:.2f}, vy={logic.vy:.2f}, "
        #      f"fuel={logic.fuel:.1f}")
        step += 1

    print(f"\nSimulation finished after {step} steps.")
    print(f"Final State: x={logic.x:.1f}, y={logic.y:.1f}, "
          f"vx={logic.vx:.2f}, vy={logic.vy:.2f}, fuel={logic.fuel:.1f}")
    print(f"Landed: {logic.landed}, Crashed: {logic.crashed}, "
          f"Success: {logic.landed_successfully}")
    print(f"Total Reward: {total_reward:.2f}")
    print("Final State Vector:", state)

File: libgame_logic/src/game_logic.cpp
/********************/
/*  game_logic.cpp  */
/*   Version 1.0    */
/*    2023/05/05    */
/********************/

#include <algorithm>
#include <cassert>
#include <cmath>
#include <iostream>
#include <numeric>
#include <stdexcept>
#include <vector>
#include "game_logic.h"

// --- Constructor ---
GameLogicCpp::GameLogicCpp(bool no_print_flag)
    : // Initialize state variables to reasonable defaults
      x(0.0), y(0.0), vx(0.0), vy(0.0), fuel(0.0), landed(false), crashed(false), landed_successfully(false),
      landing_pad_center_x(0.0), landing_pad_y(0.0), no_print(no_print_flag), last_action(0), // Default action is Noop
      // Internal state
      ax(0.0), ay(0.0), time_step(0),
      // Config defaults (will be overwritten by set_config)
      initial_x(0.0), initial_y(0.0), initial_vx(0.0), initial_vy(0.0), initial_ax(0.0), initial_ay(0.0), pcfg_g(0.0),
      lcfg_max_fuel(100.0),                       // Default value
      lpad_x1(0.0), lpad_width(0.0), pad_y1(0.0), // Landing pad
      spad_width(80.0),                           // Default starting pad width
      max_safe_vx(1.0), max_safe_vy(1.0),         // Default safe values > 0
      cfg_width(100.0), cfg_height(100.0),        // Default dimensions
      lcfg_width(10.0), lcfg_height(10.0),        // Default lander dimensions
      pcfg_mu_x(0.0), pcfg_mu_y(0.0), gcfg_terrain_y(0.0), ground_level(0.0)
{
    // reset(); // Reset is called after set_config typically
}

// --- Configuration Setting ---
void GameLogicCpp::set_config(double cfg_w, double cfg_h, double gcfg_pad_y1, double gcfg_terrain_y_val,
                              double gcfg_max_v_x, double gcfg_max_v_y, double pcfg_gravity, double pcfg_fric_x,
                              double pcfg_fric_y, double lcfg_w, double lcfg_h, double lcfg_fuel,
                              double gcfg_spad_width, double gcfg_lpad_width, const std::vector<double>& gcfg_x0_vec,
                              const std::vector<double>& gcfg_v0_vec, const std::vector<double>& gcfg_a0_vec)
{
    if (gcfg_x0_vec.size() != 2 || gcfg_v0_vec.size() != 2 || gcfg_a0_vec.size() != 2)
    {
        throw std::runtime_error("GameLogicCpp::set_config: Initial state vectors must have size 2.");
    }

    cfg_width      = cfg_w;
    cfg_height     = cfg_h;
    pad_y1         = gcfg_pad_y1;
    gcfg_terrain_y = gcfg_terrain_y_val;

    // Store absolute values for max speeds
    max_safe_vx    = std::abs(gcfg_max_v_x);
    max_safe_vy    = std::abs(gcfg_max_v_y);
    // Prevent division by zero in get_state
    if (max_safe_vx == 0.0) max_safe_vx = 1e-6;
    if (max_safe_vy == 0.0) max_safe_vy = 1e-6;

    pcfg_g        = pcfg_gravity;
    pcfg_mu_x     = pcfg_fric_x;
    pcfg_mu_y     = pcfg_fric_y;

    lcfg_width    = lcfg_w;
    lcfg_height   = lcfg_h;
    lcfg_max_fuel = lcfg_fuel;
    spad_width    = gcfg_spad_width;
    lpad_width    = gcfg_lpad_width;

    initial_x     = gcfg_x0_vec[0];
    initial_y     = gcfg_x0_vec[1];
    initial_vx    = gcfg_v0_vec[0];
    initial_vy    = gcfg_v0_vec[1];
    initial_ax    = gcfg_a0_vec[0];
    initial_ay    = gcfg_a0_vec[1];

    // Initial pad position might need to be set here if not always dynamic

    recalculate_derived_values(); // Calculate ground level etc.
}

void GameLogicCpp::update_pad_positions(double spad_x1, double lpad_x1_new)
{
    // Assuming spad_x1 is not used based on Python logic, but keeping param
    // Assuming lpad_width is fixed and set during set_config or has a default
    // If lpad_width needs to be passed too, modify the signature
    this->lpad_x1 = lpad_x1_new;
    (void)spad_x1;
    // Need lpad_width to be set before calling this
    recalculate_derived_values();
}

void GameLogicCpp::recalculate_derived_values()
{
    // Calculate landing pad center and y-coordinate
    // Ensure lpad_width is valid before calculation
    if (cfg_width > 0 && cfg_height > 0)
    { // Basic check
        landing_pad_center_x = lpad_x1 + lpad_width / 2.0;
        landing_pad_y        = pad_y1; // Pad Y seems fixed relative to terrain/config

        // Calculate ground level based on current config
        ground_level         = cfg_height - lcfg_height - gcfg_terrain_y;
    }
    else
    {
        // Handle error or set defaults if config is invalid
        landing_pad_center_x = 0.0;
        landing_pad_y        = 0.0;
        ground_level         = 0.0;
        if (!no_print)
        {
            std::cerr << "Warning: Invalid config dimensions in recalculate_derived_values." << std::endl;
        }
    }
    // Ensure max fuel is positive
    if (lcfg_max_fuel <= 0) lcfg_max_fuel = 1e-6;
}

void GameLogicCpp::reset()
{
    x                   = initial_x;
    y                   = initial_y;
    vx                  = initial_vx;
    vy                  = initial_vy;
    ax                  = initial_ax;
    ay                  = initial_ay + pcfg_g; // Apply gravity to initial vertical acceleration
    fuel                = lcfg_max_fuel;
    landed              = false;
    crashed             = false;
    landed_successfully = false;
    time_step           = 0;
    last_action         = 0; // Reset last action
}

// Overload reset to accept specific pad positions and calculate initial x
void GameLogicCpp::reset(double spad_x1, double lpad_x1_new)
{
    update_pad_positions(spad_x1, lpad_x1_new); // Update landing pad info first

    // Call the standard reset logic to reset most variables (vy, fuel, flags etc.)
    reset();

    // Calculate the lander's starting x to be centered on the start pad
    // This mirrors the logic in Python's reset_pad_positions
    x = spad_x1 + (spad_width / 2.0) - (lcfg_width / 2.0);
    // Keep the y position from the standard reset (initial_y)
}

void GameLogicCpp::apply_action(int action)
{
    last_action = action;

    if (fuel <= 0.0)
    {
        return; // No thrust if out of fuel
    }

    switch (action)
    {
    case 1:        // Thrust Up
        vy -= 0.3; // Instantaneous change, adjust as needed (Matches Python)
        fuel -= 1.0;
        break;
    case 2: // Thrust Left
        vx -= 0.05;
        fuel -= 0.5;
        break;
    case 3: // Thrust Right
        vx += 0.05;
        fuel -= 0.5;
        break;
    case 0: // Noop
    default:
        // Do nothing
        break;
    }

    // Clamp fuel to zero
    fuel = std::max(0.0, fuel);
}

void GameLogicCpp::update_physics()
{
    // Update velocity with acceleration (gravity is already in ay)
    // vx += ax; // No horizontal acceleration
    vy += ay;

    // Apply friction/air resistance
    vx *= (1.0 - pcfg_mu_x);
    vy *= (1.0 - pcfg_mu_y);

    // Update position
    x += vx;
    y += vy;

    // Ground collision is handled in check_landing_crash
}

void GameLogicCpp::check_landing_crash()
{
    // Only check for landing/crash after the first time step
    if (time_step == 0) { return; }

    // Check if the lander's bottom edge is at or below ground level
    if (y >= ground_level)
    {
        y                   = ground_level; // Ensure it rests exactly on the ground
        landed              = true;

        // Check if horizontally within the landing pad boundaries
        bool on_landing_pad = (x >= lpad_x1 && (x + lcfg_width) <= (lpad_x1 + lpad_width));

        // Check if speeds are within safe limits
        bool safe_speed     = (std::abs(vx) < max_safe_vx && std::abs(vy) < max_safe_vy);

        if (on_landing_pad && safe_speed)
        {
            landed_successfully = true;
            vx                  = 0.0; // Stop movement on successful landing
            vy                  = 0.0;
            if (!no_print) { std::cout << "GameLogicCpp: Successful landing!" << std::endl; }
        }
        else
        {
            crashed             = true;
            landed_successfully = false;
            vx                  = 0.0; // Stop movement on crash
            vy                  = 0.0;
            if (!no_print)
            {
                if (!on_landing_pad) { std::cout << "GameLogicCpp: Crashed - outside landing area!" << std::endl; }
                if (std::abs(vx) >= max_safe_vx)
                {
                    std::cout << "GameLogicCpp: Crashed - excessive horizontal speed! (|";
                    std::cout << vx << "| >= " << max_safe_vx << ")" << std::endl;
                }
                if (std::abs(vy) >= max_safe_vy)
                {
                    std::cout << "GameLogicCpp: Crashed - excessive vertical speed! (|";
                    std::cout << vy << "| >= " << max_safe_vy << ")" << std::endl;
                }
            }
        }
    }
}

std::pair<std::vector<double>, bool> GameLogicCpp::update(int action)
{
    if (is_done())
    {
        // Return current state and True for done
        return {get_state(), true};
    }

    apply_action(action);
    update_physics();
    check_landing_crash();

    time_step++;

    bool done                 = is_done();
    std::vector<double> state = get_state();

    return {state, done};
}

bool GameLogicCpp::is_done() const
{
    // Episode ends if landed or crashed
    return landed || crashed;
}

std::vector<double> GameLogicCpp::get_state() const
{
    // Calculate distance to target landing pad center
    double dist_target_x      = x - landing_pad_center_x;
    // Target Y is the top of the pad (landing_pad_y)
    // Distance is current Y relative to pad Y. Positive means above.
    double dist_target_y      = y - landing_pad_y;

    // Normalize the state for the NN
    // Ensure denominators are not zero (handled in set_config/recalculate)
    std::vector<double> state = {
        vx / max_safe_vx,           // Normalized Vx
        vy / max_safe_vy,           // Normalized Vy
        dist_target_x / cfg_width,  // Normalized distance X
        dist_target_y / cfg_height, // Normalized distance Y (relative to screen height)
        fuel / lcfg_max_fuel        // Normalized Fuel
    };

    // Apply clipping similar to Python
    state[0] = std::clamp(state[0], -2.0, 2.0); // vx_norm
    state[1] = std::clamp(state[1], -2.0, 2.0); // vy_norm
    state[2] = std::clamp(state[2], -5.0, 5.0); // dist_x_norm
    state[3] = std::clamp(state[3], -5.0, 5.0); // dist_y_norm
    // Fuel is already clamped between 0 and 1 implicitly by normalization

    return state;
}

// New get_state overload: Writes directly to the provided buffer
void GameLogicCpp::get_state(double* pOutputs, size_t outputsSize) const
{

    assert(outputsSize >= 5 && "GameLogicCpp::get_state: Output buffer size must be at least 5.");
    (void)outputsSize;

    // Calculate distance to target landing pad center
    double dist_target_x = x - landing_pad_center_x;
    double dist_target_y = y - landing_pad_y;

    // Calculate normalized values directly into the output buffer
    // Ensure denominators are not zero (handled in set_config/recalculate)
    pOutputs[0]          = vx / max_safe_vx;           // Normalized Vx
    pOutputs[1]          = vy / max_safe_vy;           // Normalized Vy
    pOutputs[2]          = dist_target_x / cfg_width;  // Normalized distance X
    pOutputs[3]          = dist_target_y / cfg_height; // Normalized distance Y
    pOutputs[4]          = fuel / lcfg_max_fuel;       // Normalized Fuel

    // Apply clipping directly to the buffer elements
    pOutputs[0]          = std::clamp(pOutputs[0], -2.0, 2.0); // vx_norm
    pOutputs[1]          = std::clamp(pOutputs[1], -2.0, 2.0); // vy_norm
    pOutputs[2]          = std::clamp(pOutputs[2], -5.0, 5.0); // dist_x_norm
    pOutputs[3]          = std::clamp(pOutputs[3], -5.0, 5.0); // dist_y_norm
}

// New update overload: Writes state to buffer, returns done flag
bool GameLogicCpp::update(int action, double* pStateOutput, size_t stateOutputSize)
{
    if (is_done())
    {
        // Populate the output buffer with the current state and return true
        get_state(pStateOutput, stateOutputSize); // Use the new get_state overload
        return true;
    }

    apply_action(action);
    update_physics();
    check_landing_crash(); // Handle ground collision and state changes

    time_step++;

    bool done = is_done();
    // Populate the output buffer with the new state
    get_state(pStateOutput, stateOutputSize);

    return done; // Return only the done flag
}
